test/module/libs/common/combine_latest_until_first_completed_test.cpp

```diff
@@ -0,0 +1,243 @@
+
+  IntervalObservableHelper(time_point start, duration period)
+      : observable_(rxcpp::observable<>::interval(start, period)),
+        start_(start),
+        period_(period) {}
+
+  virtual ~IntervalObservableHelper() = default;
+
+  long valueAtTime(time_point time) const {
+    return 1 + (time - start_) / period_;
+  }
+
+  long lastValue() const {
+    return last_emitted_value;
+  }
+
+  virtual rxcpp::observable<long> getObservable() {
+    return rememberLastVal(observable_);
+  }
+
+  rxcpp::observable<long> rememberLastVal(rxcpp::observable<long> observable) {
+    return observable.tap([this](long val) { this->last_emitted_value = val; });
+  }
+
+  rxcpp::observable<long> observable_;
+  time_point start_;
+  duration period_;
+  long last_emitted_value{0};
+};
+
+struct IntervalObservableTakeNHelper : public IntervalObservableHelper {
+  IntervalObservableTakeNHelper(duration period, size_t taken)
+      : IntervalObservableHelper(period), taken_(taken) {}
+
+  IntervalObservableTakeNHelper(time_point start, duration period, size_t taken)
+      : IntervalObservableHelper(start, period), taken_(taken) {}
+
+  rxcpp::observable<long> getObservable() override {
+    return rememberLastVal(observable_.take(taken_));
+  }
+
+  size_t taken_;
+};
+
+using IntervalObservableHelperVariant =
+    boost::variant<IntervalObservableHelper, IntervalObservableTakeNHelper>;
+
+template <typename... Types>
+void checkValues(
+    const std::vector<IntervalObservableHelperVariant> &observable_helpers,
+    size_t current_observable_idx,
+    bool &completed,
+    long current_test_value,
+    Types... args) {
+  if (current_observable_idx == 0) {
+    EXPECT_FALSE(completed) << "Got values when already completed!";
+  }
+  checkValues(observable_helpers,
+              current_observable_idx,
+              completed,
+              current_test_value);
+  checkValues(
+      observable_helpers, current_observable_idx + 1, completed, args...);
+}
+
+template <>
+void checkValues(
+    const std::vector<IntervalObservableHelperVariant> &observable_helpers,
+    size_t current_observable_idx,
+    bool &completed,
+    long current_test_value) {
+  auto last_value =
+      iroha::visit_in_place(observable_helpers[current_observable_idx],
+                            [](const auto &observable_helper) {
+                              return observable_helper.lastValue();
+                            });
+  EXPECT_EQ(last_value, current_test_value)
+      << "Got wrong value from observable #" << current_observable_idx;
+
+  iroha::visit_in_place(
+      observable_helpers[current_observable_idx],
+      [=, &completed](const IntervalObservableTakeNHelper &observable_helper) {
+        const long completion_value = observable_helper.taken_;
+        EXPECT_LE(last_value, completion_value)
+            << "Got value after the last one from observable #"
+            << current_observable_idx;
+        completed |= completion_value <= last_value;
```

Comment:
I would change the order of arguments, e.g. `last_value >= completion_value`, so that they are in the same order as in statement above.